### DISCUSSION TITLE
Greedy lock on the migrations table

### DISCUSSION THREAD

**Discussion by gvko** [sequelize/umzug#144] [Ideas] [answered]
Does Umzug or Sequelize do a lock on the migrations table (the table in which the executed migrations are stored, f.e. `SequelizeMeta`) before the migrations are to be executed and until they are finished?

If so, is it a greedy (no read/write possibilities) or a lazy lock?

I went through both the Umzug and Sequelize Migrations documentation and could not find information about that.

**Comment by jukkah** [comment]
No, currently it's not locking the table at all.

**Comment by caseyduquettesc** [comment]
Would help with #59

**Comment by jaulz** [comment]
Any chance to implement it? Are you willing to accept a PR?

**Comment by atorma** [comment]
I've used this: https://github.com/SGrondin/umzug-sync

**Comment by mmkal** [comment]
Paging @gvko @jukkah @caseyduquettesc @jaulz @atorma @SGrondin on this (three-year-) old issue! Do you still use umzug? Would you still be interested in a built-in locking feature? If so, are there any existing solutions you could share to use as inspiration/for feature discovery? Is umzug-sync still a good starting point for building something more general/database-agnostic?

Edit: also paging the 👍 ers - @DanielHreben, @ningshen, @bogdan, @pmoleri, and @ghohmann-asapp

**Reply by caseyduquettesc** [reply]
We decided to run the migrations before the application is started as part of the deployment process. It adds downtime, but it's minor and just an internal app.

**Reply by gvko** [reply]
Hey @mmkal , thanks for reaching out! 🙌 
I've been using **umzug** this whole time and am still using it, yes.

I've built 2 separate solutions so far:
1. smth similar to @caseyduquettesc - running a service that executes migrations before rolling out the actual applicaiton update. That requires infra changes on the CI/CD pipeline, so it's not an umzug-only solution.
2. recently I implemented a locking mechanism on my own that has been running successfully in our Production environment with 5 replicas of the same service, being deployed at the same time. This can be an umzug-only solution. I can submit a PR, if you'd like to look at it.

**Reply by mmkal** [reply]
@gvko it'd be great to see your solution. Might make sense to wait a few days though, I'm planning on making the event emitter support async handlers, which will make locking easier to implement and more reliable.

It'd also be good to know if you use the sequelize storage shipped with umzug. Locking varies a lot by database dialect, so it'd be non-trivial to build a generalised one.

**Reply by gvko** [reply]
@mmkal , yeah, I can wait even a few weeks. I'm not in a hurry. 😆 

I'm using the sequelize instance, but executing a raw SQL query in order to acquire/release the lock.
I'll create it in a separate repo of mine this weekend and I'll link it here.
Cuz currently it's not usable standalone. It's tightly integrated with the project I'm using it in rn, so I gotta make it more generic. 👍

**Reply by bmxpiku** [reply]
I actually used this adjusted version of this https://github.com/DanielHreben/sequelize-migration-lock/blob/master/index.js 
which works pretty well with AWS feature where instance while provisioning new updated ones, gets a little bit delayed, so it's just enough time to roll out a migration.

**Reply by EdwardIII** [reply]
@gvko Is this where you implemented your locking solution? https://github.com/gvko/feathers-migrations

**Reply by gvko** [reply]
@EdwardIII , no, I haven't made it public yet. I kinda' forgot about this thread. I'll be looking into this in the following weekend(s) and update here. :)

**Reply by EdwardIII** [reply]
@gvko that would be amazing! We're upgrading to sequelize 6, which somehow causes umzug-sync to break with these errors (pasted below). the migrations run fine themselves with just umzug, it's just with umzug-sync they're bust.

Any chance of a sneaky gist preview (if the main work is being done in a single file that can easily be pasted into a gist?)

```
Executing (c443e1b5-38a6-4c3b-8a44-3df01ac338b1): START TRANSACTION;
Executing (c443e1b5-38a6-4c3b-8a44-3df01ac338b1): SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;
Executing (c443e1b5-38a6-4c3b-8a44-3df01ac338b1): ROLLBACK;
(node:35) UnhandledPromiseRejectionWarning: Error: rollback has been called on this transaction(c443e1b5-38a6-4c3b-8a44-3df01ac338b1), you can no longer use it. (The rejected query is attached as the 'sql' property of this error)
    at checkTransaction (/usr/src/app/server/node_modules/sequelize/lib/sequelize.js:598:23)
    at /usr/src/app/server/node_modules/sequelize/lib/sequelize.js:618:9
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at emitUnhandledRejectionWarning (internal/process/promises.js:168:15)
    at processPromiseRejections (internal/process/promises.js:247:11)
    at processTicksAndRejections (internal/process/task_queues.js:96:32)
(node:35) Error: rollback has been called on this transaction(c443e1b5-38a6-4c3b-8a44-3df01ac338b1), you can no longer use it. (The rejected query is attached as the 'sql' property of this error)
    at checkTransaction (/usr/src/app/server/node_modules/sequelize/lib/sequelize.js:598:23)
    at /usr/src/app/server/node_modules/sequelize/lib/sequelize.js:618:9
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

**Comment by SamSpiri** [comment]
Hi, dear @gvko bless you. Please show what you have on the topic. "Just do it."

**Comment by gvko** [comment, marked as answer]
Alright, guys, finally it's here: [db-migrator-distributed](https://www.npmjs.com/package/db-migrator-distributed). For the lack of a better name, that's what I chose... :D 

Enjoy, give feedback and contribute! 🙌 
cc @SamSpiri @EdwardIII @mmkal @jaulz

**Reply by SamSpiri** [reply]
This is awesome! Will try, thanks in advance.

**Reply by SamSpiri** [reply]
Should this be mentioned in the upstream docs probably? ;)

**Reply by gvko** [reply]
My pleasure! 🙌 
Any feedback and contributions are more than welcome!

**Reply by EdwardIII** [reply]
Interesting @gvko we ended up going for an approach similar to the one you originally linked to. Cool to see your implementation!
